File: core/src/main/java/com/cangwang/core/cwmodule/ex/ModuleExManager.java
package com.cangwang.core.cwmodule.ex;

import android.content.res.Configuration;
import android.os.Handler;
import android.os.Looper;
import android.support.v4.util.ArrayMap;
import android.util.Log;

import com.cangwang.core.cwmodule.api.BackPressStack;
import com.cangwang.core.cwmodule.api.ModuleBackpress;
import com.cangwang.core.util.ModuleUtil;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/**
 * Created by cangwang on 2016/12/26.
 */

public class ModuleExManager {
    private List<String> modules = new ArrayList<>();   //模块名字
    protected ArrayMap<String,CWAbsExModule> allModules = new ArrayMap<>();   //模块实体
    private String template;

    public List<String> getModuleNames(){
        return modules;
    }

    public void moduleConfig(List<String> modules) {
        this.modules = modules;
    }

    public void moduleConfig(String template){
        this.template = template;
    }

    public String getTemplate(){
        return template;
    }

    private Handler handler;
    private ExecutorService pool;

    public Handler getHandler(){
        if (handler == null){
            handler = new Handler(Looper.getMainLooper());
        }
        return handler;
    }

    public ExecutorService getPool(){
        if (pool ==null){
            pool = Executors.newSingleThreadExecutor();
        }
        return pool;
    }

    public CWAbsExModule getModuleByNames(String name){
        if (!ModuleUtil.empty(allModules))
            return allModules.get(name);
        return null;
    }

    public void remove(String name){
        if (!ModuleUtil.empty(allModules)){
            allModules.remove(name);
        }
    }

    public void putModule(String name,CWAbsExModule module){
        allModules.put(name,module);
    }

    public void onResume(){
        for (CWAbsExModule module:allModules.values())
            if (module !=null)
                module.onResume();
    }

    public void onPause(){
        for (CWAbsExModule module:allModules.values())
            if (module !=null)
                module.onPause();
    }

    public void onStop(){
        for (CWAbsExModule module:allModules.values())
            if (module !=null)
                module.onStop();
    }

    public void onConfigurationChanged(Configuration newConfig){
        for (CWAbsExModule module:allModules.values())
            if (module!=null)
                module.onOrientationChanges(newConfig.orientation == Configuration.ORIENTATION_LANDSCAPE);
    }

    public void onDestroy(){
        handler = null;
        pool=null;
        for (CWAbsExModule module:allModules.values()) {
            if (module != null) {
                module.onDestroy();
            }
        }
    }

    public boolean onBackPressed(){
        boolean hasCallback =false;
//        for (CWAbsExModule module:allModules.values()) {
//            if (module != null) {
//                boolean isCall=module.onBackPress();
//                if (isCall)
//                    hasCallback =true;
//            }
//        }
        if (BackPressStack.getInstance().getStack().size() >0){
            Log.e("ModuleExManager","peek "+BackPressStack.getInstance().getStack().toString());
            hasCallback =BackPressStack.getInstance().getStack().peek().onBackPress();
        }
        return hasCallback;
    }
}

File: core/src/main/java/com/cangwang/core/cwmodule/ex/ModuleExManager.kt
package com.cangwang.core.cwmodule.ex

import com.cangwang.core.cwmodule.api.BackPressStack
import android.os.Looper
import android.content.res.Configuration
import android.os.Handler
import android.util.Log
import androidx.collection.ArrayMap
import com.cangwang.core.util.ModuleUtil
import java.util.ArrayList
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

/**
 * Created by cangwang on 2016/12/26.
 */
class ModuleExManager {
    var moduleNames: List<String> = ArrayList() //模块名字
        private set
    var allModules = ArrayMap<String?, CWAbsExModule>() //模块实体
    var template: String? = null
        private set

    fun moduleConfig(modules: List<String>) {
        moduleNames = modules
    }

    fun moduleConfig(template: String?) {
        this.template = template
    }

    private var handler: Handler? = null
    private var pool: ExecutorService? = null

    fun getHandler(): Handler {
        if (handler == null) {
            handler = Handler(Looper.getMainLooper())
        }
        return handler!!
    }

    fun getPool(): ExecutorService {
        if (pool == null) {
            pool = Executors.newSingleThreadExecutor()
        }
        return pool!!
    }

    fun getModuleByNames(name: String?): CWAbsExModule? {
        return if (!ModuleUtil.empty(allModules)) allModules[name] else null
    }

    fun remove(name: String?) {
        if (!ModuleUtil.empty(allModules)) {
            allModules.remove(name)
        }
    }

    fun putModule(name: String?, module: CWAbsExModule) {
        allModules[name] = module
    }

    fun onResume() {
        for (module in allModules.values) module?.onResume()
    }

    fun onPause() {
        for (module in allModules.values) module?.onPause()
    }

    fun onStop() {
        for (module in allModules.values) module?.onStop()
    }

    fun onConfigurationChanged(newConfig: Configuration) {
        for (module in allModules.values) module?.onOrientationChanges(newConfig.orientation == Configuration.ORIENTATION_LANDSCAPE)
    }

    fun onDestroy() {
        handler = null
        pool = null
        for (module in allModules.values) {
            module?.onDestroy()
        }
    }

    fun onBackPressed(): Boolean {
        var hasCallback = false
        //        for (CWAbsExModule module:allModules.values()) {
//            if (module != null) {
//                boolean isCall=module.onBackPress();
//                if (isCall)
//                    hasCallback =true;
//            }
//        }
        if (BackPressStack.stack.size > 0) {
            Log.e("ModuleExManager", "peek " + BackPressStack.stack.toString())
            hasCallback = BackPressStack.stack.peek().onBackPress()
        }
        return hasCallback
    }
}